### PR TITLE
Parse uint64 to int64

### PR DESCRIPTION
**Pull request by sparrc**

### CODE REVIEW

File: models/points_test.go [hunk 1]
```diff
@@ -1509,3 +1509,51 @@ func TestPrecisionString(t *testing.T) {
...
+}
+
+func TestFieldsMarshalBinaryOverMaxInt64(t *testing.T) {
+	f := models.Fields{
+		"value":  uint64(1000),
+		"value2": uint64(9223372036854775808),
+	}
+	b := f.MarshalBinary()
+
+	exp := []byte(`value=1000i,value2=`)
```
Comment (sparrc) [inline models/points_test.go]: This is what it means to "treat it as nil"

**Comment by pauldix** [comment]
Hmmm, I would argue that unsupported types should all return an error. `MarshalBinary` should be updated to return `([]byte, error)`.

Automatically converting types has caused a bunch of confusion with people using the library. If I'm reading this correctly, the reason you're adding `uint64` here to begin with is that they're getting written as strings and that's messing people up.

Better to just have the list of types we support and throw errors for everything else.

**Comment by jwilder** [comment]
@pauldix @sparrc  If we want to return an error for `uint64`, it needs to be higher up in `point` code than `MarshalBinary` I think.  The reason `MarshalBinary` does not return an error is that there is an invariant within the `point` struct that all `fields` it references are able to be marshaled.  Instead of making `MarshalBinary` return an error (which will ripple as a big change everywhere), I'd suggest making `NewPoint` and `ParsePoint` return an error instead so you can't create a point with an invalid field type.

From an end user's perspective, not handling `uint64` is kind of annoying.  We cast the other `uint` types to a `int64` so it's odd that we don't allow `uint64` even though we can store values that are less than `math.MaxInt64` without issue.  I think it would be better to allow `uint64` and have the value overflow if you write something larger than `math.MaxInt64`.  Alternatively, allow `uint64` but return an error on overflow.  With the latter, we'd have to make sure write endpoints don't start failing whole batches of writes because one point is invalid.

**Comment by pauldix** [comment]
If the user is passing in a `uint64` then they're working in a typed language. Otherwise they're trying to parse a value for which `int64` would overflow. In this latter case, we should return an error on parse.

In the former, it makes sense to force them to use a type that we actually support. Thus they can make the explicit conversion from `uint64` to `int64` and deal with the consequences themselves.

**Comment by jwilder** [comment]
@pauldix With that reasoning, then we should disallow all integer types other than `int64` as well.  Just to frame this, this issue is only for the Go client.  Other languages would get an error parsing the line protocol because the value exceeds the `int64` range.  We lose whatever type they stored when it's serialized to a string over the wire.

My main issue with the `uint64` case is consistency.  Disallowing it makes it a _special_ type that you have to handle as a one-off everywhere.  That is annoying as a go client end-user right now because you can add anything as a field since it's just an `interface{}`.

**Comment by jwilder** [comment]
I'm fine with being explicit that the other `uint` types would return an error as well.  That would be consistent as a client API.

**Comment by pauldix** [comment]
I think there are two issues here that need to be handled separately. First, parsing the line protocol. I think that if the value exceeds what can be represented as an `int64` we should be throwing an error. I don't like the idea of silently overflowing the value. It's better for the user to get a response back that says there was an error.

The second issue is about the Go client and how it behaves. I see the point that it takes an interface, but the truth is that we really only support 4 underlying types. I would prefer to only support those and return an error if the type is anything other than those. That way we avoid doing automatic conversion or casting that may not be obvious to the user.

Let them determine how to manage types that we don't support.

We could refine the client so that a point has field values added to it and there's a different method for each type. e.g.

```go
point.SetInt64Field(fieldName string, value int64)
point.SetFloat64Field(fieldName string, float64 int64)
point.SetBoolField(fieldName string, value bool)
point.SetStringField(fieldName string, value string)
```

Although I think that approach is kind of ugly, it would limit what people do with the Go client and make things much more clear.

**Comment by corylanou** [comment]
Based on the benchmarks I saw, we could just reflect to detect the type and return an error if it wasn't a supported type with virtually no more overhead.  A few more benchmarks to test this theory would be good though.

**Comment by jwilder** [comment]
@pauldix The line protocol parser returns and error for integer outside the `int64` range currently so that case is already handled.

For go client, I'm good with returning an error for all unsupported field types and forcing users to make the conversion.  I think we can just validate the fields in `NewPoint` and return an error as opposed to adding settings though.  I'd rather we move towards a model where `Points` are immutable once created personally.

**Comment by pauldix** [comment]
@jwilder sounds good +1

**Comment by sparrc** [comment]
I still find it confusing that we support all primitive data-types except `uint64`, but should I close this PR?

In my opinion, rolling over the value is a behavior that is completely understandable, and I think anyone who is writing software will also understand. There is also the fact that you can query the value back and recast to `uint64` to get the original (non-negative) value.

FWIW: MySQL and MongoDB also just roll-over `uint64` types

**Comment by otoolep** [comment]
@sparrc -- by rolling-over, I presume you mean the conversion to int64 (casting). "Rolling over" sounds like it is related to overflow.

**Comment by sparrc** [comment]
@otoolep correct, casting to int64. I was saying "rolling over" because if uint64 is above the max int64 value, it will overflow to a negative int64

**Comment by sparrc** [comment]
I believe we've decided not to do this

**Pull request closed without merging**